FreeRTOS-Plus/Source/FreeRTOS-IoT-Libraries-LTS-Beta2/c_sdk/platform/freertos/transport/src/tls_freertos.c

```diff
@@ -0,0 +1,611 @@
+        }
+        else
+        {
+            LogError( ( "(Network connection %p) Failed to send TLS close-notify: mbedTLSError= %s : %s.",
+                        pNetworkContext,
+                        mbedtlsHighLevelCodeOrDefault( tlsStatus ),
+                        mbedtlsLowLevelCodeOrDefault( tlsStatus ) ) );
+        }
+    }
+    else
+    {
+        /* WANT_READ and WANT_WRITE can be ignored. Logging for debugging purposes. */
+        LogInfo( ( "(Network connection %p) TLS close-notify sent; ",
+                   "received %s as the TLS status can be ignored for close-notify."
+                   ( tlsStatus == MBEDTLS_ERR_SSL_WANT_READ ) ? "WANT_READ" : "WANT_WRITE",
+                   pNetworkContext ) );
+    }
+
+    /* Call socket shutdown function to close connection. */
+    Sockets_Disconnect( pNetworkContext->tcpSocket );
+
+    /* Free mbed TLS contexts. */
+    sslContextFree( &( pNetworkContext->sslContext ) );
+
+    /* Free the contexts for random number generation. */
+    mbedtls_ctr_drbg_free( &ctrDrgbContext );
+    mbedtls_entropy_free( &entropyContext );
+
+    /* Clear the mutex functions for mbed TLS thread safety. */
+    mbedtls_threading_free_alt();
+}
+
+/*-----------------------------------------------------------*/
+
+int32_t TLS_FreeRTOS_recv( NetworkContext_t * pNetworkContext,
+                           void * pBuffer,
+                           size_t bytesToRecv )
+{
+    int32_t tlsStatus = 0;
+
+    tlsStatus = ( int32_t ) mbedtls_ssl_read( &( pNetworkContext->sslContext.context ),
+                                              pBuffer,
+                                              bytesToRecv );
+
+    if( ( tlsStatus == MBEDTLS_ERR_SSL_TIMEOUT ) ||
+        ( tlsStatus == MBEDTLS_ERR_SSL_WANT_READ ) ||
+        ( tlsStatus == MBEDTLS_ERR_SSL_WANT_WRITE ) )
+    {
+        /* Mark these set of errors as a timeout for the libraries to retry
+         * the read. */
+        tlsStatus = 0;
+    }
+    else if( tlsStatus < 0 )
+    {
+        LogError( ( "Failed to read data: errorStatus=%d", tlsStatus ) );
+    }
+    else
+    {
+        /* Empty else marker. */
+    }
+
+    return tlsStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+int32_t TLS_FreeRTOS_send( NetworkContext_t * pNetworkContext,
+                           const void * pBuffer,
+                           size_t bytesToSend )
+{
+    int32_t tlsStatus = 0;
+
+    tlsStatus = ( int32_t ) mbedtls_ssl_write( &( pNetworkContext->sslContext.context ),
+                                               pBuffer,
+                                               bytesToSend );
+
+    if( ( tlsStatus == MBEDTLS_ERR_SSL_TIMEOUT ) ||
+        ( tlsStatus == MBEDTLS_ERR_SSL_WANT_READ ) ||
+        ( tlsStatus == MBEDTLS_ERR_SSL_WANT_WRITE ) )
+    {
+        /* Mark these set of errors as a timeout for the libraries to retry
+         * the send. */
+        tlsStatus = 0;
```

Comment:
If we are marking this as MBEDTLS_ERR_SSL_TIMEOUT it would be better to use that instead of the magic nuber 0 here, at least that is what the comment is making me believe, but I suspect that the constant is a negative number. 

It seems that the intent is that we decide here which errors are recoverable and we should retry and which ones not, and we assume that the caller will retry when we return 0, which implies that this code depends on the caller to behave in a specific way - which should be documented in the function header or here in the comments

Reply:
Replied in the earlier comment.